File: Loenn/entities/dashSpring.lua
local utils = require("utils")

local springDepth = -8501
local springTexture = "objects/BrokemiaHelper/dashSpring/00"

local springUp = {}

springUp.name = "BrokemiaHelper/dashSpring"
springUp.depth = springDepth
springUp.justification = {0.5, 1.0}
springUp.texture = springTexture
springUp.selection = function(room, entity)
    return utils.rectangle(entity.x - 6, entity.y - 4, 12, 4)
end
springUp.placements = {
    name = "up",
    data = {
        spritePath = "objects/BrokemiaHelper/dashSpring/",
        playerCanUse = true,
        ignoreHoldables = false,
        ignoreRedBoosters = false
    }
}

local springDown = {}

springDown.name = "BrokemiaHelper/dashSpringDown"
springDown.depth = springDepth
springDown.justification = {0.5, 1.0}
springDown.texture = springTexture
springDown.rotation = math.pi
springDown.selection = function(room, entity)
    return utils.rectangle(entity.x - 6, entity.y, 12, 4)
end
springDown.placements = {
    name = "down",
    data = {
        spritePath = "objects/BrokemiaHelper/dashSpring/",
        playerCanUse = true,
        ignoreHoldables = false,
        ignoreRedBoosters = false
    }
}

local springLeft = {}

springLeft.name = "BrokemiaHelper/wallDashSpringLeft"
springLeft.depth = springDepth
springLeft.justification = {0.5, 1.0}
springLeft.texture = springTexture
springLeft.rotation = math.pi / 2
springLeft.selection = function(room, entity)
    return utils.rectangle(entity.x, entity.y - 6, 4, 12)
end
springLeft.placements = {
    name = "left",
    data = {
        spritePath = "objects/BrokemiaHelper/dashSpring/",
        playerCanUse = true,
        ignoreHoldables = false,
        ignoreRedBoosters = false
    }
}

local springRight = {}

springRight.name = "BrokemiaHelper/wallDashSpringRight"
springRight.depth = springDepth
springRight.justification = {0.5, 1.0}
springRight.texture = springTexture
springRight.rotation = -math.pi / 2
springRight.selection = function(room, entity)
    return utils.rectangle(entity.x - 4, entity.y - 6, 4, 12)
end
springRight.placements = {
    name = "right",
    data = {
        spritePath = "objects/BrokemiaHelper/dashSpring/",
        playerCanUse = true,
        ignoreHoldables = false,
        ignoreRedBoosters = false
    }
}

return {
    springUp,
    springDown,
    springRight,
    springLeft
}

File: DashSpring.cs
using Celeste;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;
using MonoMod.Utils;
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Runtime.CompilerServices;

namespace BrokemiaHelper {
    [CustomEntity("BrokemiaHelper/dashSpring", "BrokemiaHelper/wallDashSpringRight", "BrokemiaHelper/wallDashSpringLeft", "BrokemiaHelper/dashSpringDown")]
    public class DashSpring : Spring {

        private static FieldInfo playerCanUseInfo = typeof(Spring).GetField("playerCanUse", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.GetField);
        private static FieldInfo spriteInfo = typeof(Spring).GetField("sprite", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.GetField);

        private static MethodInfo BounceAnimateInfo = typeof(Spring).GetMethod("BounceAnimate", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.InvokeMethod);

        private bool ignoreRedBoosters;

        public DashSpring(Vector2 position, Orientations orientation, string spritePath, bool playerCanUse, bool ignoreHoldables, bool ignoreRedBubble)
            : base(position, orientation == (Orientations)3 ? Orientations.Floor : orientation, playerCanUse) {
            Orientation = orientation;
            ignoreRedBoosters = ignoreRedBubble;
            DynData<Spring> selfData = new DynData<Spring>(this);

            if(ignoreHoldables) {
                Remove(Get<HoldableCollider>());
            }

            // Only one other player collider is added so it can easily be removed
            Remove(Get<PlayerCollider>());
            Add(new PlayerCollider(OnCollide));
            Sprite sprite = (Sprite)spriteInfo.GetValue(this);
            sprite.Reset(GFX.Game, spritePath);
            sprite.Add("idle", "", 0f, default(int));
            sprite.Add("bounce", "", 0.07f, "idle", 0, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 4, 5);
            sprite.Add("disabled", "white", 0.07f);
            sprite.Play("idle");
            sprite.Origin.X = sprite.Width / 2f;
            sprite.Origin.Y = sprite.Height;
            if (orientation == (Orientations)3) {
                selfData.Get<StaticMover>("staticMover").SolidChecker = ((Solid s) => CollideCheck(s, Position - Vector2.UnitY));
                selfData.Get<StaticMover>("staticMover").JumpThruChecker = ((JumpThru jt) => CollideCheck(jt, Position - Vector2.UnitY));
                Collider = new Hitbox(16f, 6f, -8f, 0f);
                Get<PufferCollider>().Collider = new Hitbox(16f, 10f, -8f, 0f);
                sprite.Rotation = (float)Math.PI;
            }

        }

        public DashSpring(EntityData data, Vector2 offset)
            : this(data.Position + offset, GetOrientationFromName(data.Name), data.Attr("spritePath", "objects/BrokemiaHelper/dashSpring/"), data.Bool("playerCanUse", true), data.Bool("ignoreHoldables", false), data.Bool("ignoreRedBoosters", false)) {
        }

        public static Orientations GetOrientationFromName(string name) {
            switch (name) {
                case "BrokemiaHelper/dashSpring":
                    return Orientations.Floor;
                case "BrokemiaHelper/wallDashSpringRight":
                    return Orientations.WallRight;
                case "BrokemiaHelper/wallDashSpringLeft":
                    return Orientations.WallLeft;
                case "BrokemiaHelper/dashSpringDown":
                    return (Orientations)3;
                default:
                    throw new Exception("Dash Spring name doesn't correlate to a valid Orientation!");
            }
        }

        protected new void OnCollide(Player player) {
            if (player.StateMachine.State == 9 || !(bool)playerCanUseInfo.GetValue(this) || !player.DashAttacking) {
                return;
            }
            if(ignoreRedBoosters && player.StateMachine.State == Player.StRedDash) {
                return;
            }
            if (Orientation == Orientations.Floor) {
                if (player.Speed.Y >= 0f) {
                    BounceAnimateInfo.Invoke(this, null);
                    player.SuperBounce(Top);
                }
                return;
            }
            if (Orientation == Orientations.WallLeft) {
                if (player.SideBounce(1, Right, CenterY)) {
                    BounceAnimateInfo.Invoke(this, null);
                }
                return;
            }
            if (Orientation == Orientations.WallRight) {
                if (player.SideBounce(-1, Left, CenterY)) {
                    BounceAnimateInfo.Invoke(this, null);
                }
                return;
            }
            if (Orientation == (Orientations)3) {
                if (player.Speed.Y <= 0f) {
                    BounceAnimateInfo.Invoke(this, null);
                    player.SuperBounce(Bottom + player.Height);
                    
                    DynData<Player> playerData = new DynData<Player>(player);
                    playerData["varJumpSpeed"] = player.Speed.Y = 185f;
                    SceneAs<Level>().DirectionalShake(Vector2.UnitY, 0.1f);
                }
                return;
            }
            throw new Exception("Orientation not supported!");
        }
    }
}
